Add support for sending grpc server backend metrics via ORCA

twthorn opened the pull request:
<!--
  Thank you for your contribution to the Vitess project.
  How to contribute: https://vitess.io/docs/contributing/
  Please first make sure there is an open Issue to discuss the feature/fix suggested in this PR.
  If this is a new feature, please mark the Issue as "RFC".
 -->

<!-- if this PR is Work in Progress please create it as a Draft Pull Request -->

## Description

This is a [supported feature of GRPC](https://grpc.io/docs/guides/custom-backend-metrics/) for better load balancing by grpc client (can receive the metrics and use them to pick the correct grpc server). It is opt-in so its backward compatible (existing clients will ignore the metrics). Once this is merged, clients can be updated to use the metrics for better load balancing (both at connection initialization, and for long-running connections that are on overloaded servers, to find an underutilized server)

As far as I can tell, [no one else](https://github.com/search?q=ServerMetricsRecorder+-path%3A*_test.go+path%3A*.go+-path%3A*test*+&type=code) has done this on open source GitHub. There are cases of people forking grpc-go or setting static values (eg in tests), but none that dynamically read cpu/memory and support cross platform. As a later iteration, I may try to make a go package that gives these basic functionalities, then Vitess would only need to import the package and call the top level get CPU / Mem and the package handles all the rest (cgroup, host-level, windows, linux, mac etc.)

## Related Issue(s)

https://github.com/vitessio/vitess/issues/18270

## Checklist

Will update docs once we finalize the PR.

-   [ ] "Backport to:" labels have been added if this change should be back-ported to release branches
-   [ ] If this change is to be back-ported to previous releases, a justification is included in the PR description
-   [ ] Tests were added or are not required
-   [ ] Did the new or modified tests pass consistently locally and on CI?
-   [ ] Documentation was added or is not required

## Deployment Notes

<!-- Notes regarding deployment of the contained body of work. These should note any db migrations, etc. -->

dbussink commented on go/vt/servenv/grpc_server.go:
```diff
@@ -325,6 +350,22 @@ func serveGRPC() {
+	if err != nil {
+		return 0
+	}
+	return vmStat.UsedPercent
```
Does this logic all work properly inside containers where there's allocated memory budgets etc? We can't assume all memory is available for the given process here and we need to make sure we honor things like cgroup memory allocations.

dbussink reviewed (commented):
I don't like it, but I think we need a flag for this. At least for a while so that in case people run into trouble with this new feature, they can turn it off. 

We should then first set it disabled by default so people can opt in, then if it works well for folks we can turn the default to on in a later version and if the feedback is that there's no downsides, we can potentially remove the flag again.

twthorn commented:
@dbussink thank you for the review. That makes sense to wrap behind a flag initially, updated based on this feedback

twthorn commented:
@dbussink can you take another look when you get the chance? Thanks!

twthorn commented on go/vt/servenv/grpc_server.go:
```diff
@@ -327,6 +352,82 @@ func serveGRPC() {
+		return -1
+	}
+
+	time.Sleep(interval)
```
Good point, will update in a moment

twthorn commented:
@dbussink Thanks for the feedback. I updated it so we won't be sleeping in application code, and instead compare to the previously recorded sample time. In tests we do a sleep to get a reasonably accurate value.

Also added a docker portion to test cgroups v2 with go linux base image. This can also be expanded to test other platforms like windows, mac, etc.

twthorn commented on go/vt/servenv/metrics_linux.go:
```diff
@@ -0,0 +1,36 @@
+
+func getHostMemory() (float64, error) {
+	return getHostMemoryUsage()
+}
```
Nice catch, yes this can be simplified, updating

twthorn commented:
As far as I can tell this vdiff2 check is flakey and failing on other PRs. I believe it safe to ignore.

dbussink commented on go/vt/vtgate/vstream_manager_test.go:
```diff
@@ -526,7 +526,7 @@ func TestVStreamMulti(t *testing.T) {
 	ctx, cancel := context.WithCancel(context.Background())
 	defer cancel()
-	cell := "aa"
+	cell := "ab"
```
Any reason for the changes in this file?

twthorn replied:
I saw this was one of the flakey tests failing in unit tests. I contributed this test a while back, and noticed if we use unique cells, there's no risk of parallel tests reading each other's metrics (reduces flakiness). I'll add a comment to make this clear
